app/routes.py

```diff
@@ -1,2 +1,191 @@
+
+@task_bp.route("", methods=["POST", "GET"])
+def handle_tasks():
+    if request.method == "POST":
+        request_body = request.get_json()
+        if "title" not in request_body or "description" not in request_body or "completed_at" not in request_body:
+            return jsonify({"details": "Invalid data"}), 400
+        new_task = Task(
+            title= request_body["title"],
+            description= request_body["description"],
+            completed_at = request_body["completed_at"]
+        )
+        db.session.add(new_task)
+        db.session.commit()
+        if new_task.completed_at:
+            new_task.is_complete = True
+        
+        response = {}
+        response["task"] = new_task.to_dict()
+        return jsonify(response), 201
+
+    elif request.method == "GET":
+        sort_query = request.args.get("sort")
+        if sort_query:
+            if sort_query == "asc":
+                tasks = Task.query.order_by(Task.title).all()
+            elif sort_query == "desc":
+                tasks = Task.query.order_by(Task.title.desc()).all()
+        else: 
+            tasks = Task.query.all()
+        
+        response_body = []
+        for task in tasks:
+            response_body.append(task.to_dict())
+        return jsonify(response_body)
+
+@task_bp.route("/<task_id>", methods=["GET", "PUT", "DELETE"])
+def handle_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response("", 404)
+
+    if request.method == "GET":
+        response = {}
+        response["task"] = task.to_dict()
+        if task.goal_id:
+            response["task"]["goal_id"]=task.goal_id
+        return jsonify(response)
+
+    elif request.method == "PUT":
+        request_body = request.get_json()
+        task.title = request_body["title"]
+        task.description = request_body["description"]
+        if "completed_at" in request_body:
+            task.completed_at = request_body["completed_at"]
+            task.is_complete = True
+        db.session.commit()
+
+        response_body = {}
+        response_body["task"] = task.to_dict()
+        return jsonify(response_body)
+
+    elif request.method == "DELETE":
+        db.session.delete(task)
+        db.session.commit()
+        
+        response_body = {"details": f"Task {task.id} \"{task.title}\" successfully deleted"}
+        return jsonify(response_body)
+
+@task_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_complete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response("", 404)
+    task.is_complete = True
+    db.session.commit()
+    task.completed_at = datetime.now()
+    
+    # posts to slack
+    url = 'https://slack.com/api/chat.postMessage'
+    slack_request = {
+        "token": BOT_TOKEN,
+        "channel": "task-notifications",
+        "text": f"Someone just completed the task {task.title}"
+    }
+    requests.post(url, data=slack_request)
```

Comment:
To practice the single responsibility principle, we can move the logic for the slack bot into a helper function. This will also allow the slack bot to be used by other routes too (ex: confirming post, put, or get requests, etc.)